Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/opensaml/CertificateUtil.java

```diff
@@ -76,4 +97,149 @@ private void closeStream(Closeable stream) {
             LOG.warn("Unable to close the stream {}", ex.getLocalizedMessage(), ex);
         }
     }
+
+    public static KeyStore loadKeyStore(final String storeType, final String password, final String storeLoc)
+        throws CertificateManagerException {
+        InputStream is = null;
+        KeyStore secretStore = null;
+        try {
+            secretStore = KeyStore.getInstance(storeType);
+            if (!PKCS11_TYPE.equalsIgnoreCase(storeType)) {
+                is = new FileInputStream(storeLoc);
+            }
+            secretStore.load(is, password.toCharArray());
+        } catch (final IOException | NoSuchAlgorithmException | CertificateException | KeyStoreException ex) {
+            LOG.error(KEYSTORE_ERROR_MSG, ex.getLocalizedMessage(), ex);
+            throw new CertificateManagerException(ex.getMessage(), ex);
+        } finally {
+            try {
+                if (is != null) {
+                    is.close();
+                }
+            } catch (final IOException ex) {
+                LOG.error("Unable to close keyStoreStream- KeyStoreCallbackHandler {}", ex.getLocalizedMessage(), ex);
+            }
+        }
+        return secretStore;
+    }
+
+    public static Map<String, Certificate> getKeystoreMap(final KeyStore keystore) throws KeyStoreException {
+        Map<String, Certificate> retObj = new HashMap<>();
+        Enumeration<String> alaises = keystore.aliases();
+        String alias;
+        Certificate cert;
+        String subjectKeyId;
+        while(alaises.hasMoreElements()){
+            alias = alaises.nextElement();
+            cert = keystore.getCertificate(alias);
+            subjectKeyId = getCertKeyIdSubject(cert);
+            retObj.put(subjectKeyId, cert);
+        }
+        return retObj;
+    }
+
+    public static List<Certificate> getChain(Certificate cert, final KeyStore keystore) throws KeyStoreException {
+        return getChain((X509Certificate) cert, keystore);
+    }
+    public static List<Certificate> getChain(X509Certificate cert, final KeyStore keystore) throws KeyStoreException {
+        if (null == cert) {
+            return new ArrayList<>();
+        }
+        Map<String, Certificate> keyCache = getKeystoreMap(keystore);
+        List<Certificate> chain = new ArrayList<>();
+        chain.add(cert);
+
+        Certificate certLoop = cert;
+        String aki = getCertKeyIdAuthority(certLoop); // if-root: expected-null
+        while (null != aki) {
+            certLoop = keyCache.get(aki);
+            chain.add(certLoop);
+            aki = getCertKeyIdAuthority(certLoop);
+        }
+
+        return chain;
+    }
+
+    public static String getCertKeyIdAuthority(Certificate cert) {
```

Comment:
why do we need this method?

Reply:
the subject-id and authority-id depend on one-another to figure-out the chain

Reply:
Spoke with Tran, and this is a placeholder facade for supporting multiple types of certificates in the future.